### PR TITLE
Add `WithHealthzEndpoint` as `ServeMuxOption` to register a `/healthz`endpoint

### LINKED ISSUE #2302: Health check endpoint
## 📚 Documentation

It would be great to add further documentation to https://grpc-ecosystem.github.io/grpc-gateway/docs/operations/health_check/ and add the example discussed in #1977.

Furthermore the combination of both would be nice to have one http endpoint on the gateway that calls the `check` method on the grpc server to really have a meaningful result for the health of the whole service.

If possible it would also be nice if it could be added to the options in `runtime.NewServeMux` like `WithHealthCheckEnabled(conn *grpc.ClientConn)`, to activate a "healthz" route with the described behaviour. I would guess that the need for healthchecks should be common enough for that.
If you think it shouldn't be integrated in this place, maybe there's also another place where this could fit in.

### PR DESCRIPTION
**Pull request by brumhard**
#### References to other Issues or PRs

Fixes #2302

#### Brief description of what is fixed or changed
This PR adds a `WithHealthzEndpoint` as `ServeMuxOption` to `runtime.NewServeMux()`.
It registers a `/healthz` route in the resulting `ServeMux` that forwards each request to the backing grpc server's health check endpoint.
Like this it's possible to check the whole system's health from the grpc-gateway instead of checking the gateway and grpc server on their own.
This is especially useful if you only expose the grpc-gateway.

### CODE REVIEW

File: runtime/mux_test.go [hunk 1]
```diff
@@ -466,3 +475,126 @@ func TestServeMux_HandlePath(t *testing.T) {
...
+func (g *grpcServer) Check(ctx context.Context, r *grpc_health_v1.HealthCheckRequest) (*grpc_health_v1.HealthCheckResponse, error) {
+	if g.code != codes.OK {
+		return nil, status.Error(g.code, r.GetService())
+	}
+
+	return &grpc_health_v1.HealthCheckResponse{Status: g.status}, nil
+}
+
+func (g *grpcServer) Watch(r *grpc_health_v1.HealthCheckRequest, s grpc_health_v1.Health_WatchServer) error {
+	panic("implement me")
```
Comment (johanbrandhorst) [inline runtime/mux_test.go:580]: Slightly nicer style is to use `status.Errorf`, what do you think?
```suggestion
	return status.Error(codes.Unimplemented, "unimplemented")
```

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
This is a really fantastic contribution, even includes a doc update, I'm very impressed! I just had one tiny nit for the tests but it's otherwise good to go! Make sure you run the commands in [CONTRIBUTING.md](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/CONTRIBUTING.md#i-want-to-regenerate-the-files-after-making-changes) and sign the CLA so we can merge this :).

File: runtime/mux_test.go (hunk 1 above)
Comment (brumhard) [reply runtime/mux_test.go]: Nice one, I just kept the autogenerated code for this one, but looks cleaner like this 👍

**Comment by brumhard** [comment]
Hi there. I'm really glad that you like my contribution. Sadly I'll have to wait a bit for my employer to get ready with the CLA, but I will update as soon as it works.
Also I ran the commands you mentioned and noticed that the github packages link is a bit outdated.

I would change the following
```shell
docker run -v $(pwd):/src/grpc-gateway --rm docker.pkg.github.com/grpc-ecosystem/grpc-gateway/build-env:1.16 \
    /bin/bash -c 'cd /src/grpc-gateway && \
        make install && \
        make clean && \
        make generate'
```

to 
```shell
docker run -v $(pwd):/grpc-gateway -w /grpc-gateway --rm ghcr.io/grpc-ecosystem/grpc-gateway/build-env:1.16 \
    /bin/bash -c 'make install && \
        make clean && \
        make generate'
```

and also update the link to point to [the container registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) instead of the outdated docker registry. 
Should I do this in this PR or are there reasons that you stick to the docker registry for now?

**Comment by johanbrandhorst** [comment]
> Should I do this in this PR or are there reasons that you stick to the docker registry for now?

Please make a separate PR for this, but I would much appreciate that fix!

**Comment by johanbrandhorst** [comment]
> > Should I do this in this PR or are there reasons that you stick to the docker registry for now?
> 
> Please make a separate PR for this, but I would much appreciate that fix!

Nevermind, I went and added these changes to https://github.com/grpc-ecosystem/grpc-gateway/pull/2320

**Comment by brumhard** [comment]
Hi there, sadly I'm still waiting for my company to sign the CLA. I'll update the PR asap, but not sure how long it will take.

**Comment by johanbrandhorst** [comment]
That's okay, thanks for the update!

**Comment by lovromazgon** [comment]
@brumhard any update on the CLA? I'd love to see this merged, it would be very useful!

**Comment by brumhard** [comment]
Hi @lovromazgon, sadly no. I'll ask again right after the holidays and make sure that it gets merged asap :)

**Comment by brumhard** [comment]
@googlebot I signed it!

**Comment by brumhard** [comment]
So finally the CLA is signed, just took 6 months 🕺🏻

File: runtime/mux.go [hunk 2]
```diff
@@ -204,6 +206,50 @@ func WithDisablePathLengthFallback() ServeMuxOption {
...
 }
 
+// WithHealthzEndpoint returns a ServeMuxOption that will add a /healthz endpoint to the created ServeMux.
+// When called the handler will forward the request to the upstream grpc service health check (defined in the
+// gRPC Health Checking Protocol).
+// See here https://grpc-ecosystem.github.io/grpc-gateway/docs/operations/health_check/ for more information on how
+// to setup the protocol in the grpc server.
+// If you define a service as query parameter, this will also be forwarded as service in the HealthCheckRequest.
+func WithHealthzEndpoint(conn grpc.ClientConnInterface) ServeMuxOption {
+	healthCheckClient := grpc_health_v1.NewHealthClient(conn)
```
Comment (johanbrandhorst) [inline runtime/mux.go:216]: I know I've already reviewed this, but now that I'm looking at it, wouldn't it make more sense for this parameter to be an instantiation of the health client? That way it'll be easier to test and the user can decide how to dial the client.
```suggestion
func WithHealthzEndpoint(healthCheckClient grpc_health_v1.HealthClient) ServeMuxOption {
```

Comment (brumhard) [reply runtime/mux.go]: That was a really nice hint, feels much cleaner now. I was not sure if you are using any mocking framework and couldn't find one, so I kept the fake approach.

**Review by johanbrandhorst — APPROVED** [review]
LGTM

**Comment by johanbrandhorst** [comment]
Thank you for your contribution! This is a great new feature.

**Pull request merged**